File: jflove-app/test/utils/crypto_test.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:flutter_test/flutter_test.dart';

import 'package:jflove_app/utils/crypto.dart';

/// 加密原语交叉验证测试
///
/// 验证移动端 pointycastle + x25519 加密栈与桌面端/server 端互通。
void main() {
  group('X25519 ECDH', () {
    test('密钥对生成', () {
      final kp = CryptoUtils.generateKeyPair();
      expect(kp.privateKeyRaw.length, 32);
      expect(kp.publicKeyB64.isNotEmpty, true);
      // 验证公钥是有效的 Base64
      final decoded = base64Decode(kp.publicKeyB64);
      expect(decoded.length, 32);
    });

    test('ECDH 派生 session_key 长度', () {
      // Alice 生成密钥对
      final alice = CryptoUtils.generateKeyPair();
      // Bob 生成密钥对
      final bob = CryptoUtils.generateKeyPair();

      // Alice 用自己私钥 + Bob 公钥派生
      final aliceSession = CryptoUtils.deriveSessionKey(alice.privateKeyRaw, bob.publicKeyB64);
      // Bob 用自己私钥 + Alice 公钥派生
      final bobSession = CryptoUtils.deriveSessionKey(bob.privateKeyRaw, alice.publicKeyB64);

      // 双方 session_key 应当一致（长度 32 字节）
      expect(aliceSession.length, 32);
      expect(bobSession.length, 32);
      expect(aliceSession, orderedEquals(bobSession));
    });

    test('不同密钥对派生不同 session_key', () {
      final alice = CryptoUtils.generateKeyPair();
      final bob1 = CryptoUtils.generateKeyPair();
      final bob2 = CryptoUtils.generateKeyPair();

      final session1 = CryptoUtils.deriveSessionKey(alice.privateKeyRaw, bob1.publicKeyB64);
      final session2 = CryptoUtils.deriveSessionKey(alice.privateKeyRaw, bob2.publicKeyB64);

      // 不同对端应派生出不同 session_key
      expect(session1, isNot(orderedEquals(session2)));
    });
  });

  group('ChaCha20-Poly1305 加密信封', () {
    late Uint8List sessionKey;

    setUp(() {
      final kp = CryptoUtils.generateKeyPair();
      // 用自己和自己做 ECDH（仅用于测试，不用于生产）
      sessionKey = CryptoUtils.deriveSessionKey(kp.privateKeyRaw, kp.publicKeyB64);
    });

    test('整包加密-解密往返', () {
      final plaintext = utf8.encode('{"test": "hello jflove", "number": 42}');

      final envelope = CryptoUtils.encryptEnvelope(sessionKey, Uint8List.fromList(plaintext));
      expect(envelope.nonce.isNotEmpty, true);
      expect(envelope.ciphertext.isNotEmpty, true);
      expect(envelope.nonce, isNot(equals(envelope.ciphertext)));

      final decrypted = CryptoUtils.decryptEnvelope(sessionKey, envelope.nonce, envelope.ciphertext);
      expect(utf8.decode(decrypted), equals(utf8.decode(Uint8List.fromList(plaintext))));
    });

    test('篡改密文应抛出异常', () {
      final plaintext = utf8.encode('sensitive data');
      final envelope = CryptoUtils.encryptEnvelope(sessionKey, Uint8List.fromList(plaintext));

      // 篡改密文字节
      final cipherBytes = base64Decode(envelope.ciphertext);
      cipherBytes[0] ^= 0xFF;
      final tamperedCiphertext = base64Encode(cipherBytes);

      expect(
        () => CryptoUtils.decryptEnvelope(sessionKey, envelope.nonce, tamperedCiphertext),
        throwsArgumentError,
      );
    });

    test('错误的 nonce 应抛出异常', () {
      final plaintext = utf8.encode('test data');
      final envelope = CryptoUtils.encryptEnvelope(sessionKey, Uint8List.fromList(plaintext));

      expect(
        () => CryptoUtils.decryptEnvelope(sessionKey, 'AAAAAAAAAAAAAAAAAAAAAA==', envelope.ciphertext),
        throwsArgumentError,
      );
    });
  });

  group('流式分片加密', () {
    late Uint8List sessionKey;

    setUp(() {
      final kp = CryptoUtils.generateKeyPair();
      sessionKey = CryptoUtils.deriveSessionKey(kp.privateKeyRaw, kp.publicKeyB64);
    });

    test('流式分片加密-解密往返', () {
      final plaintext = Uint8List.fromList(List.generate(1000, (i) => i % 256));

      final frame = CryptoUtils.encryptStreamChunk(sessionKey, plaintext);
      expect(frame.length > plaintext.length, true); // 有加密开销

      // 解析帧头：[4B 长度][12B nonce][密文+16B tag]
      final frameLen = (ByteData.sublistView(frame).getUint32(0, Endian.big));
      expect(frameLen, equals(frame.length - 4));

      // 提取消 nonce 以外的帧体
      final frameBody = frame.sublist(4);
      final decrypted = CryptoUtils.decryptStreamChunk(sessionKey, frameBody);
      expect(decrypted, orderedEquals(plaintext));
    });

    test('空数据加密', () {
      final plaintext = Uint8List(0);
      final frame = CryptoUtils.encryptStreamChunk(sessionKey, plaintext);
      expect(frame.length, greaterThan(4));
    });

    test('大块数据（64KB）加密往返', () {
      final plaintext = Uint8List.fromList(List.generate(64 * 1024, (i) => i % 256));
      final frame = CryptoUtils.encryptStreamChunk(sessionKey, plaintext);
      final frameBody = frame.sublist(4);
      final decrypted = CryptoUtils.decryptStreamChunk(sessionKey, frameBody);
      expect(decrypted, orderedEquals(plaintext));
    });
  });

  group('安全用例 - 加密信封往返', () {
    test('测试向量：字符串加密解密一致', () {
      final kp = CryptoUtils.generateKeyPair();
      final sessionKey = CryptoUtils.deriveSessionKey(kp.privateKeyRaw, kp.publicKeyB64);

      const testCases = [
        '{"token": "abc123"}',
        '{"detail": "操作成功"}',
        '{"files": [{"name": "test.txt", "size": 1024}]}',
      ];

      for (final tc in testCases) {
        final plaintext = utf8.encode(tc);
        final envelope = CryptoUtils.encryptEnvelope(sessionKey, Uint8List.fromList(plaintext));
        final decrypted = CryptoUtils.decryptEnvelope(sessionKey, envelope.nonce, envelope.ciphertext);
        expect(utf8.decode(decrypted), equals(tc));
      }
    });
  });
}

File: jflove-web/tests/utils/crypto.test.ts
import { describe, it, expect } from 'vitest';
import {
  arrayBufferToBase64,
  base64ToUint8Array,
  generateKeyPair,
  deriveSessionKey,
  encryptEnvelope,
  decryptEnvelope,
  generateKeyPairJS,
  deriveSessionKeyJS,
} from '../../src/utils/crypto';

describe('crypto', () => {
  it('Base64 编解码互逆', () => {
    const original = new Uint8Array([0, 1, 2, 255, 128, 64, 32, 16]);
    const b64 = arrayBufferToBase64(original);
    const decoded = base64ToUint8Array(b64);
    expect(decoded).toEqual(original);
  });

  it('生成 X25519 密钥对（Web Crypto 或纯 JS 回退）', async () => {
    const { publicKeyB64, privateKey } = await generateKeyPair();
    expect(publicKeyB64).toHaveLength(44); // 32 bytes raw → 44 base64 chars
    // 安全上下文为 CryptoKey，非安全上下文为 32B Uint8Array（回退）
    expect(privateKey).toBeTruthy();
    if (privateKey instanceof Uint8Array) {
      expect(privateKey).toHaveLength(32);
    }
  });

  it('ECDH + HKDF 派生相同的 session_key', async () => {
    const alice = await generateKeyPair();
    const bob = await generateKeyPair();

    const aliceKey = await deriveSessionKey(alice.privateKey, bob.publicKeyB64);
    const bobKey = await deriveSessionKey(bob.privateKey, alice.publicKeyB64);

    expect(aliceKey).toEqual(bobKey);
    expect(aliceKey).toHaveLength(32);
  });

  it('纯 JS 回退：X25519 + HKDF 派生相同的 session_key（HTTP 非安全上下文兼容）', () => {
    const alice = generateKeyPairJS();
    const bob = generateKeyPairJS();

    expect(alice.publicKeyB64).toHaveLength(44);
    expect(alice.privateKey).toHaveLength(32);

    const aliceKey = deriveSessionKeyJS(alice.privateKey, bob.publicKeyB64);
    const bobKey = deriveSessionKeyJS(bob.privateKey, alice.publicKeyB64);

    expect(aliceKey).toEqual(bobKey);
    expect(aliceKey).toHaveLength(32);
  });

  it('纯 JS 与 Web Crypto 跨实现互操作：session_key 一致（与后端/移动端互操作等价验证）', async () => {
    // 一端纯 JS 私钥，一端 Web Crypto 密钥，交叉 ECDH 后 HKDF 输出必须一致
    const js = generateKeyPairJS();
    const wc = await generateKeyPair();

    const keyFromJs = deriveSessionKeyJS(js.privateKey, wc.publicKeyB64);
    const keyFromWc = await deriveSessionKey(wc.privateKey, js.publicKeyB64);

    expect(keyFromJs).toEqual(keyFromWc);
    expect(keyFromJs).toHaveLength(32);
  });

  it('ChaCha20-Poly1305 加密解密互逆', async () => {
    const { privateKey } = await generateKeyPair();
    const peer = await generateKeyPair();
    const sessionKey = await deriveSessionKey(privateKey, peer.publicKeyB64);

    const plaintext = new TextEncoder().encode('Hello, JFLove!');
    const { nonce, ciphertext } = encryptEnvelope(sessionKey, plaintext);
    const decrypted = decryptEnvelope(sessionKey, nonce, ciphertext);

    expect(new TextDecoder().decode(decrypted)).toBe('Hello, JFLove!');
  });

  it('不同 session_key 解密失败', async () => {
    const alice = await generateKeyPair();
    const bob = await generateKeyPair();
    const eve = await generateKeyPair();

    const sessionKey = await deriveSessionKey(alice.privateKey, bob.publicKeyB64);
    const wrongKey = await deriveSessionKey(eve.privateKey, alice.publicKeyB64);

    const plaintext = new TextEncoder().encode('secret');
    const { nonce, ciphertext } = encryptEnvelope(sessionKey, plaintext);

    // 不同密钥解密应失败
    expect(() => decryptEnvelope(wrongKey, nonce, ciphertext)).toThrow();
  });
});
